src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorCompiler.cs

```diff
@@ -0,0 +1,140 @@
+        {
+            var item = _templateEngine.Project.GetItem(path);
+            var imports = _templateEngine.Project.FindHierarchicalItems(path, _templateEngine.Options.ImportsFileName);
+            return new CompilerCacheContext(item, imports, GetCompilationResult);
+        }
+
+        private CompilationResult GetCompilationResult(CompilerCacheContext cacheContext)
+        {
+            var projectItem = cacheContext.ProjectItem;
+            var codeDocument = _templateEngine.CreateCodeDocument(projectItem.Path);
+            var cSharpDocument = _templateEngine.GenerateCode(codeDocument);
+
+            CompilationResult compilationResult;
+            if (cSharpDocument.Diagnostics.Count > 0)
+            {
+                compilationResult = GetCompilationFailedResult(
+                    projectItem.Path,
+                    cSharpDocument.Diagnostics);
+            }
+            else
+            {
+                compilationResult = _compilationService.Compile(codeDocument, cSharpDocument);
+            }
+
+            return compilationResult;
+        }
+
+        private CompilationResult GetCompilationFailedResult(
+            string relativePath,
+            IEnumerable<RazorDiagnostic> diagnostics)
+        {
+            // If a SourceLocation does not specify a file path, assume it is produced from parsing the current file.
+            var messageGroups = diagnostics.GroupBy(razorError => razorError.Span.FilePath ?? relativePath, StringComparer.Ordinal);
+
+            var failures = new List<CompilationFailure>();
+            foreach (var group in messageGroups)
+            {
+                var filePath = group.Key;
+                var fileContent = ReadFileContentsSafely(_templateEngine.Project.GetItem(filePath));
+                var compilationFailure = new CompilationFailure(
+                    filePath,
+                    fileContent,
+                    compiledContent: string.Empty,
+                    messages: group.Select(parserError => CreateDiagnosticMessage(parserError, filePath)));
+                failures.Add(compilationFailure);
+            }
+
+            return new CompilationResult(failures);
+        }
+
+        private static DiagnosticMessage CreateDiagnosticMessage(
+            RazorDiagnostic error,
+            string filePath)
+        {
+            var sourceSpan = error.Span;
+            return new DiagnosticMessage(
+                message: error.GetMessage(),
+                formattedMessage: $"{error} ({sourceSpan.LineIndex},{sourceSpan.CharacterIndex}) {error.GetMessage()}",
+                filePath: filePath,
+                startLine: sourceSpan.LineIndex + 1,
+                startColumn: sourceSpan.CharacterIndex,
+                endLine: sourceSpan.LineIndex + 1,
+                endColumn: sourceSpan.CharacterIndex + sourceSpan.Length);
+        }
+
+        private static string ReadFileContentsSafely(RazorProjectItem projectItem)
```

Comment:
You don't need this if you already have a `RazorCodeDocument` or `RazorSourceDocument`.